hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ReplicationManager.java

```diff
@@ -478,6 +599,281 @@ private void updateInflightAction(final ContainerInfo container,
+     *source removed*     unexpected       *    expected      *
+     **********************************************************
+     *target removed*      expected        *   unexpected     *
+     **********************************************************
+     * unexpected action may happen somehow. to make it deterministic,
+     * if unexpected action happens, we just fail the completableFuture.
+     */
+
+    if (isSource && isInflightReplication) {
+      //if RM is reinitialize, inflightMove will be restored,
+      //but inflightMoveFuture not. so there will be a case that
+      //container is in inflightMove, but not in inflightMoveFuture.
+      compleleteMoveFutureWithResult(id,
+          MoveResult.UNEXPECTED_REMOVE_SOURCE_AT_INFLIGHT_REPLICATION);
+      moveScheduler.completeMove(id.getProtobuf());
+      return;
+    }
+
+    if (isTarget && !isInflightReplication) {
+      compleleteMoveFutureWithResult(id,
+          MoveResult.UNEXPECTED_REMOVE_TARGET_AT_INFLIGHT_DELETION);
+      moveScheduler.completeMove(id.getProtobuf());
+      return;
+    }
+
+    if (!(isInflightReplication && isCompleted)) {
+      if (isInflightReplication) {
+        if (isUnhealthy) {
+          compleleteMoveFutureWithResult(id,
+              MoveResult.REPLICATION_FAIL_NODE_UNHEALTHY);
+        } else if (isNotInService) {
+          compleleteMoveFutureWithResult(id,
+              MoveResult.REPLICATION_FAIL_NODE_NOT_IN_SERVICE);
+        } else {
+          compleleteMoveFutureWithResult(id,
+              MoveResult.REPLICATION_FAIL_TIME_OUT);
+        }
+      } else {
+        if (isUnhealthy) {
+          compleleteMoveFutureWithResult(id,
+              MoveResult.DELETION_FAIL_NODE_UNHEALTHY);
+        } else if (isTimeout) {
+          compleleteMoveFutureWithResult(id,
+              MoveResult.DELETION_FAIL_TIME_OUT);
+        } else if (isNotInService) {
+          compleleteMoveFutureWithResult(id,
+              MoveResult.DELETION_FAIL_NODE_NOT_IN_SERVICE);
+        } else {
+          compleleteMoveFutureWithResult(id,
+              MoveResult.COMPLETED);
+        }
+      }
+      moveScheduler.completeMove(id.getProtobuf());
+    } else {
+      deleteSrcDnForMove(container,
+          containerManager.getContainerReplicas(id));
+    }
+  }
+
+  /**
+   * add a move action for a given container.
+   *
+   * @param cid Container to move
+   * @param src source datanode
+   * @param tgt target datanode
+   */
+  public CompletableFuture<MoveResult> move(ContainerID cid,
+             DatanodeDetails src, DatanodeDetails tgt)
+      throws ContainerNotFoundException, NodeNotFoundException {
+    return move(cid, new MoveDataNodePair(src, tgt));
+  }
+
+  /**
+   * add a move action for a given container.
+   *
+   * @param cid Container to move
+   * @param mp MoveDataNodePair which contains source and target datanodes
+   */
+  public CompletableFuture<MoveResult> move(ContainerID cid,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 752-763, lines 765-775 into separate methods.